File: packages/types/src/issues/issue_worklog.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

import type { IProjectLite } from "../project";
import type { IUserLite } from "../users";

// docs/feature-specs/14-pricing-gap-remediation.md ("14a. Time Tracking and
// Work Logs", feature 1 "Saisie de temps par work item") in plane-selfhost -
// mirrors apps/api/plane/db/models/issue_worklog.py's `IssueWorklog` model.
export type TIssueWorklog = {
  id: string;
  workspace: string;
  project: string;
  issue: string;
  logged_by: string;
  logged_by_detail: IUserLite;
  duration: number;
  logged_at: string;
  description: string;
  timesheet_period: string | null;
  created_at: string;
  updated_at: string;
  created_by: string | undefined;
  updated_by: string | undefined;
};

export type TIssueWorklogMap = {
  [worklog_id: string]: TIssueWorklog;
};

export type TIssueWorklogIdMap = {
  [issue_id: string]: string[];
};

export type TWorklogOperations = {
  createWorklog: (data: Partial<TIssueWorklog>) => Promise<TIssueWorklog | undefined>;
  updateWorklog: (worklogId: string, data: Partial<TIssueWorklog>) => Promise<void>;
  removeWorklog: (worklogId: string) => Promise<void>;
};

// docs/feature-specs/14-pricing-gap-remediation.md ("14a. Time Tracking and
// Work Logs", feature 3 "Workflow d'approbation de timesheet") - mirrors
// `TimesheetPeriod`.
export type TTimesheetPeriodStatus = "draft" | "submitted" | "approved" | "rejected";

export type TTimesheetPeriod = {
  id: string;
  workspace: string;
  project: string;
  project_detail: IProjectLite;
  logged_by: string;
  logged_by_detail: IUserLite;
  period_start: string;
  period_end: string;
  status: TTimesheetPeriodStatus;
  submitted_at: string | null;
  approved_by: string | null;
  approved_by_detail: IUserLite | null;
  approved_at: string | null;
  rejection_reason: string;
  created_at: string;
  updated_at: string;
};

// docs/feature-specs/14-pricing-gap-remediation.md ("14a. Time Tracking and
// Work Logs", feature 2 "Timesheets historiques et rapports agrégés") in
// plane-selfhost - mirrors the shared response shape of
// WorkspaceWorklogReportEndpoint/ProjectWorklogReportEndpoint/
// MyWorklogReportEndpoint (apps/api/plane/app/views/issue_worklog/report.py).
export type TWorklogReportGroupBy = "project" | "member";

export type TWorklogReportRow = {
  project_id?: string;
  member_id?: string;
  total_duration: number;
  entry_count: number;
};

export type TWorklogReport = {
  group_by: TWorklogReportGroupBy;
  results: TWorklogReportRow[];
  total_duration: number;
  total_entries: number;
};

// Query filters shared by the workspace/project/"my" report endpoints and
// the CSV export trigger - a plain object rather than URLSearchParams so
// callers can build it once and pass it to both the report GET and the
// export POST.
export type TWorklogReportFilters = {
  project_id?: string[];
  member_id?: string[];
  date_from?: string;
  date_to?: string;
  group_by?: TWorklogReportGroupBy;
};

File: apps/api/plane/app/serializers/issue_worklog.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

# Python imports
from datetime import date

# Third party imports
from rest_framework import serializers

# Module imports
from .base import BaseSerializer
from .user import UserLiteSerializer
from .project import ProjectLiteSerializer
from plane.db.models import IssueWorklog, TimesheetPeriod


class IssueWorklogSerializer(BaseSerializer):
    logged_by_detail = UserLiteSerializer(read_only=True, source="logged_by")

    class Meta:
        model = IssueWorklog
        fields = [
            "id",
            "issue",
            "project",
            "workspace",
            "logged_by",
            "logged_by_detail",
            "duration",
            "logged_at",
            "description",
            "timesheet_period",
            "created_at",
            "updated_at",
            "created_by",
            "updated_by",
        ]
        read_only_fields = [
            "id",
            "issue",
            "project",
            "workspace",
            "logged_by",
            "timesheet_period",
            "created_at",
            "updated_at",
            "created_by",
            "updated_by",
        ]

    def validate_duration(self, value):
        if value < 1:
            raise serializers.ValidationError("Duration must be at least 1 minute")
        return value

    def validate_logged_at(self, value):
        if value > date.today():
            raise serializers.ValidationError("Cannot log time for a future date")
        return value


class IssueWorklogExportSerializer(BaseSerializer):
    """
    Flat, one-row-per-entry shape for the "issue_worklogs" ExporterHistory
    CSV export - see docs/feature-specs/14-pricing-gap-remediation.md
    ("14a. Time Tracking and Work Logs", feature 2, exigence 6) in
    plane-selfhost: project name, work item, author, date, duration,
    description.
    """

    project_name = serializers.CharField(source="project.name", read_only=True)
    issue_name = serializers.CharField(source="issue.name", read_only=True)
    issue_sequence_id = serializers.SerializerMethodField()
    logged_by_email = serializers.CharField(source="logged_by.email", read_only=True)
    logged_by_display_name = serializers.CharField(source="logged_by.display_name", read_only=True)

    class Meta:
        model = IssueWorklog
        fields = [
            "id",
            "project_name",
            "issue_sequence_id",
            "issue_name",
            "logged_by_email",
            "logged_by_display_name",
            "logged_at",
            "duration",
            "description",
        ]
        read_only_fields = fields

    def get_issue_sequence_id(self, obj):
        return f"{obj.project.identifier}-{obj.issue.sequence_id}"


class TimesheetPeriodSerializer(BaseSerializer):
    logged_by_detail = UserLiteSerializer(read_only=True, source="logged_by")
    approved_by_detail = UserLiteSerializer(read_only=True, source="approved_by")
    project_detail = ProjectLiteSerializer(read_only=True, source="project")

    class Meta:
        model = TimesheetPeriod
        fields = [
            "id",
            "project",
            "project_detail",
            "workspace",
            "logged_by",
            "logged_by_detail",
            "period_start",
            "period_end",
            "status",
            "submitted_at",
            "approved_by",
            "approved_by_detail",
            "approved_at",
            "rejection_reason",
            "created_at",
            "updated_at",
        ]
        read_only_fields = fields
